mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -198,13 +156,28 @@ class SessionActor(
             // Fail stream.
             originalSender ! Status.Failure(ex)
         }
+    // TODO: there is a bug. The session requests do not follow redirects
     case SessionActor.Response(originalCall, originalSender, response) =>
       logger.debug(s"Call replied with ${response.status}")
       if (response.status == StatusCodes.Unauthorized) {
         logger.info("Refreshing IAM authentication token")
 
         context.become(initializing)
-        credentialsProvider.nextToken().pipeTo(self)
+        authorization match {
+          case Some(credentialsProvider) => credentialsProvider.nextToken().pipeTo(self)
+          case None =>
+            throw new IllegalStateException("Received HTTP 401 Unauthorized but no credential provider was supplied.")
+        }
+
+        // Queue current call again.
+        self.tell(originalCall, originalSender)
+      } else if (response.status.isRedirection()) {
+        logger.debug(s"Received redirect $response")
+
+        // Reset scheduler endpoint.
+        val locationHeader = response.header[headers.Location].get
+        val redirectUri = locationHeader.uri.resolvedAgainst(requestUri)
+        context.become(initialized(redirectUri, maybeCredentials))
```

Comment:
We may want to count the redirects. It's a shame that Akka HTTP cannot follow redirects.